code/xcvm/evm/src/Router.sol

```diff
@@ -0,0 +1,149 @@
+
+    event RemoveOwners(address sender, uint256 networkId, BridgeSecurity security, address[] owners);
+
+    event Spawn(
+        bytes account,
+        uint256 networkId,
+        BridgeSecurity security,
+        bytes salt,
+        bytes spawnedProgram,
+        address[] assetAddresses,
+        uint256[] amounts
+    );
+
+    modifier onlyBridge() {
+        require(bridgesInfo[msg.sender].security != BridgeSecurity(0));
+        _;
+    }
+
+    constructor() {
+        // enable trustless bridge;
+    }
+
+    function getAsset(uint256 assetId) external returns (address) {
+        return assets[assetId];
+    }
+
+    function getBridge(uint256 networkId, BridgeSecurity security) external returns (address) {
+        return bridges[networkId][security];
+    }
+
+    function registerAsset(address assetAddress, uint128 assetId) external onlyOwner {
+        require(assetAddress != address(0), "Router: invalid address");
+        assets[assetId] = assetAddress;
+    }
+
+    function unregisterAsset(uint128 assetId) external onlyOwner {
+        delete assets[assetId];
+    }
+
+    function registerBridge(
+        address bridgeAddress,
+        BridgeSecurity security,
+        uint256 networkId
+    ) external onlyOwner {
+        require(bridges[networkId][security] == address(0), "Router: this type of bridge already registered");
+        require(bridgeAddress != address(0), "Router: invalid address");
+        require(bridgesInfo[bridgeAddress].security == BridgeSecurity(0), "Router: bridge already enabled");
+        require(security != BridgeSecurity(0), "Router: should not disable bridge while registering bridge");
+        bridgesInfo[bridgeAddress].security = security;
+        bridgesInfo[bridgeAddress].networkId = networkId;
+        bridges[networkId][security] = bridgeAddress;
+    }
+
+    function unregisterBridge(address bridgeAddress) external onlyOwner {
+        require(bridgesInfo[bridgeAddress].security != BridgeSecurity(0), "Router: bridge already disabled");
+        delete bridges[bridgesInfo[bridgeAddress].networkId][bridgesInfo[bridgeAddress].security];
+        bridgesInfo[bridgeAddress].security = BridgeSecurity(0);
+        bridgesInfo[bridgeAddress].networkId = 0;
+    }
+
+    //// TODO ? is the bridge who's gonna to provide internetwork assets transfer?
+    function _provisionAssets(
+        address payable interpreterAddress,
+        address[] memory erc20AssetList,
+        uint256[] memory amounts
+    ) internal {
+        require(
+            erc20AssetList.length == amounts.length,
+            "Router: asset list size should be equal to amount list size"
+        );
+        if (msg.value > 0) {
+            bool sent = interpreterAddress.send(msg.value);
+            require(sent, "Failed to send Ether");
+        }
+        for (uint256 i = 0; i < erc20AssetList.length; i++) {
+            IERC20(erc20AssetList[i]).transferFrom(msg.sender, interpreterAddress, amounts[i]);
```

Comment:
Unless I am missing something, this code assume that the ETH will be sent to this contract while the ERC tokens kept on the sender account, probably nice to explain that (its some sort of gas optimisation right, unless allowing this contract to transfer on behalf of the bridge is not more efficient than a normal transfer ,in which case its probably better to do a 2 step transfer here: relay -> bridge -> interpreter)